nifi-commons/nifi-record-path/src/main/java/org/apache/nifi/record/path/RecordFieldRemover.java

```diff
@@ -0,0 +1,169 @@
+            schema.removePath(path);
+        }
+    }
+
+    private void addToPathIfNotRoot(FieldValue field, RecordFieldRemovalPath path) {
+        if (field.getParent().isPresent()) {
+            path.add(field.getField().getFieldName());
+        }
+    }
+
+    private static class RecordPath {
+        private String recordPath;
+
+        public RecordPath(final String recordPath) {
+            this.recordPath = preprocessRecordPath(recordPath);
+        }
+
+        public boolean isAppliedToAllElementsInCollection() {
+            return recordPath.endsWith("[*]") || recordPath.endsWith("[0..-1]");
+        }
+
+        @Override
+        public String toString() {
+            return recordPath;
+        }
+
+        private String preprocessRecordPath(final String recordPath) {
+            if (recordPath.endsWith("]")) {
+                return unifyRecordPathEnd(recordPath);
+            }
+            return recordPath;
+        }
+
+        private String unifyRecordPathEnd(final String recordPath) {
+            String lastSquareBracketsOperator = getLastSquareBracketsOperator(recordPath);
+            if (lastSquareBracketsOperator.equals("[*]")) {
+                return recordPath.substring(0, recordPath.lastIndexOf('[')) + "[*]";
+            } else if (lastSquareBracketsOperator.equals("[0..-1]")) {
+                return recordPath.substring(0, recordPath.lastIndexOf('[')) + "[0..-1]";
+            } else {
+                return recordPath;
+            }
+        }
+
+        private String getLastSquareBracketsOperator(final String recordPath) {
+            int beginIndex = recordPath.lastIndexOf('[');
+            return recordPath.substring(beginIndex).replaceAll("\\s","");
+        }
+
+        public boolean pathRemovalRequiresSchemaModification() {
+            return allSquareBracketsContainAsteriskOnly(recordPath);
+        }
+
+        private boolean allSquareBracketsContainAsteriskOnly(String recordPath) {
+            boolean allSquareBracketsContainAsteriskOnly = true;
+            boolean inSquareBrackets = false;
+            for (int i = 0; i < recordPath.length() && allSquareBracketsContainAsteriskOnly; ++i) {
+                char character = recordPath.charAt(i);
+                if (inSquareBrackets) {
+                    switch (character) {
+                        case ' ':
+                        case '*':
+                            break;
+                        case ']':
+                            inSquareBrackets = false;
+                            break;
+                        default:
+                            allSquareBracketsContainAsteriskOnly = false;
+                    }
+                } else {
+                    if (character == '[') {
+                        inSquareBrackets = true;
+                    }
+                }
+            }
+            return allSquareBracketsContainAsteriskOnly;
+        }
+    }
```

Comment:
This could be significantly simplified:

```suggestion
    private boolean isAppliedToAllElementsInCollection(String recordPath) {
        // ends with [*] or [0..-1]
        return recordPath.matches(".*\\[\\s*\\*\\s*\\]$") || recordPath.matches(".*\\[\\s*0\\s*\\.\\.\\s*\\-1\\s*\\]$");
    }

    private boolean pathRemovalRequiresSchemaModification(String recordPath) {
        boolean pathNotReferencesIndividualArrayElements = !recordPath.matches(".*\\[(\\s*\\d+(\\s*,\\s*)?)+\\].*");
        boolean pathNotReferencesIndividualMapElements = !recordPath.matches(".*\\[\\s*'.*'\\s*\\].*");

        return pathNotReferencesIndividualArrayElements && pathNotReferencesIndividualMapElements;
    }
```